### PR TITLE
Add Memoization For Tails

### PR DESCRIPTION
**Pull request by adamnew123456**
This changeset makes `Stream` memoize its tail - rather than calling the tail function more than once, it will only be called once and its value cached.

### CODE REVIEW

**Comment by adamnew123456** [comment]
Sorry about the forest of minor fix commits - I really shouldn't be doing this past midnight. It works, though.

The one thing that I don't quite understand is why `drop` and `dropWhile` were using newly created streams - I simply rewrote those to return `self` and I don't think they function any differently.

**Comment by dotnetCarpenter** [comment]
I'm just guessing here but perhaps `drop` and `dropWhile` returns new streams to adhere to the immutable principle? @tech-no-crat

**Comment by adamnew123456** [comment]
I reverted `drop` and `dropWhile` to creating a copy of the stream - it just required a few minor changes to the `Stream` constructor.

**Comment by dionyziz** [comment]
This patch assumes that streams are generally deterministic, which may not always be the case. Making them deterministic is a valid opinionated change, but let's discuss this before merging. Don't we want to take into account the possibility of a tail changing between calls?

File: lib/index.js [hunk 1]
```diff
@@ -208,16 +230,15 @@ Stream.prototype = {
 
           self = self.tail();
         }
-        
-        // create clone/a contructor which accepts a stream?
-        return new Stream( self.headValue, self.tailPromise );
+
+        return new Stream(self.headValue, self.tailPromise);
```
Comment (dionyziz) [inline lib/index.js]: Please maintain the coding style of the original if possible :)

**Comment by dotnetCarpenter** [comment]
@adamnew123456 just want to add; thanks for writing unit tests that actually improve the overall code coverage!

**Comment by adamnew123456** [comment]
> Please maintain the coding style of the original if possible for consistency :)

Thanks for the heads up - that should be fixed in the latest commit, let me know if I missed anything else.

> This patch assumes that streams are generally deterministic, which may not always be the case. Making them deterministic is a valid opinionated change, but let's discuss this before merging. Don't we want to take into account the possibility of a tail changing between calls?

That's a good point - I suppose that there are two ways to go about this:

1. Leave Stream alone and let it work the same way, while adding a LazyStream 
   (or CachingStream) as a persistent version of Stream.
2. Make Stream persistent, and then have some kind of out for people who need
   a non-deterministic stream.

Option one should be easy, and people dependant upon Stream's current 
non-determinism don't have to change their code, but the trade off is that 
people who *can* use a LazyStream but don't modify their code don't gain 
anything.

Option two would require some more work in the case of non-deterministic 
streams, more than it would in option one. On the other hand, option two is a 
win for people who use Stream deterministically - they'll see some speedup 
(how fast depends upon the complexity of the function that you use to 
generate your tails) since tails are cached rather than being recomputed.

I think it comes down to how many people are actually using Streams 
non-deterministically, and how much. My first guess would be *not many*
and *not much,* but I don't really know.

When it comes to use cases, my thought is this - I think having a value which
can change *merely by looking at it* is totally insane. I use Streams as a 
way of performing work when you need it, rather than as a kind of list which 
can change its length and contents without warning (although they definitely can
do that, I just don't use them that way). I'm want to hear an advocate of the 
other side and see their use cases, though.

**Comment by dionyziz** [comment]
I prefer to go with option two. Then we can have an EagerStream which evaluates every time?

**Comment by adamnew123456** [comment]
Sure - I should have something soon.

**Comment by adamnew123456** [comment]
There were a couple of changes to the external API that should be made clear (they're in the commit message, but I want to make sure everyone sees them):

- `Stream.Eager` is the constructor now used for eager streams
- `Stream.iterateEager` and `Stream.continuallyEager` are the eager version of their lazy counterparts. I thought that it made sense to have both versions since they generate their tails directly from some function.

**Comment by dionyziz** [comment]
Great, thanks!

**Pull request merged**